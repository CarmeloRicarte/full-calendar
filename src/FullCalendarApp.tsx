export const FullCalendarApp = () => {
  return <div>FullCalendarApp</div>;
};
