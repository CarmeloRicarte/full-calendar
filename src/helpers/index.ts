export * from "./CalendarLocalizer";
export * from "./getCalendarMessages";
