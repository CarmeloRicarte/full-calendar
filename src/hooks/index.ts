export * from "./dispatchSelector";
export * from "./useCalendarStore";
export * from "./useForm";
export * from "./useUiStore";
