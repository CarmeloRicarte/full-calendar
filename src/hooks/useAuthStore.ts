import { useSelector } from "react-redux";
import { calendarApi } from "../api";
import {
  clearErrorMessage,
  onChecking,
  onLogin,
  onLogout,
  RootState,
} from "../store";
import { useAppDispatch } from "./dispatchSelector";

export const useAuthStore = () => {
  const { status, user, errorMessage } = useSelector(
    (state: RootState) => state.auth
  );
  const dispatch = useAppDispatch();

  const startLogin = async ({
    email,
    password,
  }: {
    email: string;
    password: string;
  }) => {
    dispatch(onChecking());
    try {
      const { data } = await calendarApi.post("/auth", {
        email,
        password,
      });
      setTokenLocalStorage(data.token);

      dispatch(
        onLogin({
          name: data.name,
          uid: data.uid,
        })
      );
    } catch (error) {
      dispatch(onLogout("Credenciales incorrectas"));
      setTimeout(() => {
        dispatch(clearErrorMessage());
      }, 1000);
    }
  };

  const startRegister = async ({
    name,
    email,
    password,
  }: {
    name: string;
    email: string;
    password: string;
  }) => {
    dispatch(onChecking());
    try {
      const { data } = await calendarApi.post("/auth/new", {
        name,
        email,
        password,
      });
      setTokenLocalStorage(data.token);

      dispatch(
        onLogin({
          name: data.name,
          uid: data.uid,
        })
      );
    } catch (error: any) {
      if (error.response.data?.msg) {
        dispatch(onLogout(error.response.data?.msg));
      } else {
        const keyFirstElement = Object.keys(error.response.data?.errors)[0];
        keyFirstElement &&
          dispatch(onLogout(error.response.data?.errors[keyFirstElement]?.msg));
      }

      setTimeout(() => {
        dispatch(clearErrorMessage());
      }, 1000);
    }
  };

  const setTokenLocalStorage = (token: string) => {
    localStorage.setItem("token", token);
    localStorage.setItem(
      "token-expires-date",
      new Date().getTime() + 2 * 60 * 60 * 1000 + ""
    );
  };

  const checkAuthToken = async () => {
    const token = localStorage.getItem("token");
    if (!token) return dispatch(onLogout());
    try {
      const { data } = await calendarApi.get("/auth/renew");

      // logout user if token expired
      const expirationTokenDate = localStorage.getItem("token-expires-date");
      if (expirationTokenDate && +expirationTokenDate < new Date().getTime()) {
        dispatch(onLogout());
      } else {
        setTokenLocalStorage(data);
        dispatch(
          onLogin({
            name: data.name,
            uid: data.uid,
          })
        );
      }
    } catch (error) {
      localStorage.clear();
      dispatch(onLogout());
    }
  };

  return {
    //* Properties
    status,
    user,
    errorMessage,
    //* Methods
    startLogin,
    startRegister,
    checkAuthToken,
  };
};
