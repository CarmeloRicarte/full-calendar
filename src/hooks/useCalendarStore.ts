import { useAppDispatch } from "./dispatchSelector";
import { useSelector } from "react-redux";
import { onSetActiveEvent, RootState } from "../store";
import { CalendarEvent } from "../calendar/interfaces";

export const useCalendarStore = () => {
  const dispatch = useAppDispatch();
  const { events, activeEvent } = useSelector(
    (state: RootState) => state.calendar
  );

  const setActiveEvent = (calendarEvent: CalendarEvent) => {
    dispatch(onSetActiveEvent(calendarEvent));
  };

  return {
    //* Properties
    events,
    activeEvent,
    //*Methods
    setActiveEvent,
  };
};
