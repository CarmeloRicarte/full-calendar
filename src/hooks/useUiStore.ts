import { useSelector } from "react-redux";
import { onCloseDateModal, onOpenDateModal, RootState } from "../store";
import { useAppDispatch } from "./dispatchSelector";

export const useUiStore = () => {
  const dispatch = useAppDispatch();
  const { isDateModalOpen } = useSelector((state: RootState) => state.ui);

  const openDateModal = () => {
    dispatch(onOpenDateModal());
  };

  const closeDateModal = () => dispatch(onCloseDateModal());

  return {
    //* Properties
    isDateModalOpen,

    //*Methods
    openDateModal,
    closeDateModal,
  };
};
