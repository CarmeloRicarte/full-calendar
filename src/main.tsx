import React from "react";
import ReactDOM from "react-dom/client";
import { BrowserRouter } from "react-router-dom";
import { FullCalendarApp } from "./FullCalendarApp";
import "./styles.css";

ReactDOM.createRoot(document.getElementById("root") as HTMLElement).render(
  <React.StrictMode>
    <BrowserRouter>
      <FullCalendarApp />
    </BrowserRouter>
  </React.StrictMode>
);
