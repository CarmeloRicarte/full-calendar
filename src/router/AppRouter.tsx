import { useEffect } from "react";
import { Navigate, Route, Routes } from "react-router-dom";
import { LoginPage } from "../auth/pages";
import { CalendarPage } from "../calendar";
import { useAuthStore } from "../hooks";
import { AUTH_STATUS } from "../store";

export const AppRouter = () => {
  const { checkAuthToken, status } = useAuthStore();

  useEffect(() => {
    checkAuthToken();
  }, []);

  if (status === AUTH_STATUS.CHECKING) {
    return <h3>Loading...</h3>;
  } else {
  }

  return (
    <Routes>
      {status !== AUTH_STATUS.AUTHENTICATED ? (
        <Route path="/auth/*" element={<LoginPage />} />
      ) : (
        <Route path="/*" element={<CalendarPage />} />
      )}

      <Route path="/*" element={<Navigate to="/auth/login" />} />
    </Routes>
  );
};
