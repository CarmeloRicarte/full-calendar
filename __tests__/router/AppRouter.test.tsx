import { describe, test, expect, beforeEach, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import React from "react";
import { MemoryRouter } from "react-router-dom";
import * as useAuthStore from "../../src/hooks/useAuthStore";
import { AUTH_STATUS } from "../../src/store";
import { AppRouter } from "../../src/router/AppRouter";

// mock useAuthStore
vi.mock("../../src/hooks/useAuthStore");

// mock CalendarPage. Is not mandatory to create a mock
//of all component for check if app have navigated to CalendarPage
vi.mock("../../src/calendar", () => ({
  CalendarPage: () => <h1>CalendarPage</h1>,
}));

describe("Tests of <AppRouter />", () => {
  const mockCheckAuthToken = vi.fn();

  beforeEach(() => {
    vi.clearAllMocks();
  });

  test("should show loading", () => {
    (useAuthStore as any).useAuthStore = vi.fn().mockReturnValue({
      status: AUTH_STATUS.CHECKING,
      checkAuthToken: mockCheckAuthToken,
    });
    render(<AppRouter />);
    expect(mockCheckAuthToken).toHaveBeenCalled();
    expect(screen.getByText("Loading...")).toBeTruthy();
  });

  test("should show login if user is not authenticated", () => {
    (useAuthStore as any).useAuthStore = vi.fn().mockReturnValue({
      status: AUTH_STATUS.NOT_AUTHENTICATED,
      checkAuthToken: mockCheckAuthToken,
    });
    const { container } = render(
      <MemoryRouter>
        <AppRouter />
      </MemoryRouter>
    );
    expect(screen.getByText("Ingreso")).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  test("should show calendar page if user is authenticated", () => {
    (useAuthStore as any).useAuthStore = vi.fn().mockReturnValue({
      status: AUTH_STATUS.AUTHENTICATED,
      checkAuthToken: mockCheckAuthToken,
    });

    render(
      <MemoryRouter>
        <AppRouter />
      </MemoryRouter>
    );

    expect(screen.getByText("CalendarPage")).toBeTruthy();
  });
});
