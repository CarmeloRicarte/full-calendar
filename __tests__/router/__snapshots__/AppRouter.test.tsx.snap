// Vitest Snapshot v1

exports[`Tests of <AppRouter /> > should show login if user is not authenticated 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Ingreso
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              data-testid="login-email-input"
              name="loginEmail"
              placeholder="Correo"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              data-testid="login-password-input"
              name="loginPassword"
              placeholder="Contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid gap-2 mt-3"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Login"
            />
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Registro
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              data-testid="register-name-input"
              name="registerName"
              placeholder="Nombre"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              data-testid="register-email-input"
              name="registerEmail"
              placeholder="Correo"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              data-testid="register-password-input"
              name="registerPassword"
              placeholder="Contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              data-testid="register-password2-input"
              name="registerPassword2"
              placeholder="Repita la contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid gap-2 mt-3"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Crear cuenta"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
